# Tabla Market Review y algoritmos RandomForest


###  **1. Justificación de los algoritmos elegidos**

El problema de predicción planteado en este TFM requiere estimar el comportamiento futuro de las ventas de vehículos y clasificar el desempeño de las marcas automotrices. Para ello se seleccionaron dos variantes del algoritmo Random Forest:

- **RandomForestClassifier** → para clasificar marcas según su comportamiento (crecimiento, estabilidad o caída).  
- **RandomForestRegressor** → para predecir valores continuos como unidades futuras y variaciones proyectadas.

La elección de Random Forest se fundamenta en las siguientes razones técnicas:

### **1.1. Manejo de relaciones no lineales**
Las ventas de vehículos presentan fluctuaciones, estacionalidad, variaciones abruptas y comportamientos no lineales. Random Forest es especialmente adecuado para capturar estas relaciones complejas sin necesidad de supuestos estadísticos rígidos.

### **1.2. Robustez frente a ruido y variabilidad**
El datasetcontiene más de **307.000 registros** con variabilidad por marca, año y segmento. Random Forest tolera:

- ruido,  
- valores atípicos,  
- datos faltantes,  
- distribuciones heterogéneas.

### **1.3. Capacidad para trabajar con múltiples variables derivadas**
El modelo utiliza variables como:

- crecimiento interanual,acumulados, variación porcentual, unidades históricas, ratios derivados.
Random Forest maneja sin problemas variables numéricas y categóricas transformadas.

### **1.4. Interpretación de importancia de variables**
Una ventaja clave para el TFM es que Random Forest permite identificar:
- qué variables explican mejor las ventas,  
- qué factores influyen en el crecimiento,  
- qué indicadores son más relevantes para cada marca.

Esto aporta valor estratégico al proyecto 100PREORIGINALES.

### **1.5. Estabilidad y compatibilidad técnica**
Random Forest funciona de manera estable en:
- Windows, Anaconda, Python puro,
sin requerir librerías de optimización complejas (BLAS, MKL), lo que evita errores de ejecución y facilita la reproducibilidad.

---

### **2. Proceso de entrenamiento de los modelos**

El entrenamiento se realizó siguiendo un pipeline estructurado:

####  **2.1. División del dataset**
Se utilizó una partición estándar:

- **70% entrenamiento**  
- **30% prueba**

Garantizando que los datos de prueba no fueran utilizados en el entrenamiento.

### # **2.2. Preparación de variables**
Se aplicaron:

- normalización numérica,  
- codificación de categorías,  
- generación de variables derivadas,  
- eliminación de columnas irrelevantes.

####  **2.3. Entrenamiento del modelo**
Se entrenaron:

- **RandomForestClassifier(n_estimators=200, max_depth=None)**  
- **RandomForestRegressor(n_estimators=300, max_depth=None)**

Los modelos se ajustaron para capturar patrones históricos sin sobreajustar.

### **3. Ajuste de hiperparámetros**

Se aplicará un ajuste básico de hiperparámetros mediante:

- **GridSearchCV**,  
- **RandomizedSearchCV**,  
- pruebas manuales controladas.


###  **4. Validación de los modelos**

Se aplicaron dos métodos de validación:

### #**4.1. Validación Hold-Out**
Comparación entre:
- datos de entrenamiento,  
- datos de prueba.

### **4.2. Validación cruzada (Cross-Validation)**
Se utilizó **K-Fold (k=5)** para evaluar estabilidad del modelo.

Esto permitió verificar:

- consistencia de métricas,  
- ausencia de sobreajuste,  
- comportamiento estable en diferentes particiones.

---

### **5. Evaluación mediante métricas apropiadas**

#### **5.1. RandomForestClassifier**
Se utiliza métricas estándar para clasificación:

- **Precision**  
- **Recall**  
- **F1-score**  
- **Accuracy**  
- **Matriz de confusión**

### **5.2. RandomForestRegressor**
Se utiliza métricas para regresión:

- **MAE (Mean Absolute Error)**  
- **RMSE (Root Mean Squared Error)**  
- **R² (Coeficiente de determinación)**


#### Conexión a vista market_review_simplificado y limpiando data

In [12]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("postgresql+psycopg2://postgres:ASUS@localhost:5433/postgres")

query = """
SELECT anio, mes, marca, familia, segmento, unidades
FROM spo.market_review_simplificado
WHERE marca = 'CHEVROLET'
"""

df = pd.read_sql(query, engine)
df['fecha'] = pd.to_datetime(df['anio'].astype(str) + '-' + df['mes'].astype(str) + '-01')
df = df.sort_values('fecha')

print(df.head())
print(df.isna().sum())


       anio  mes      marca    familia   segmento  unidades      fecha
46306  2007    1  CHEVROLET      CORSA  AUTOMOVIL         1 2007-01-01
35046  2007    1  CHEVROLET       AVEO  AUTOMOVIL        63 2007-01-01
46381  2007    1  CHEVROLET     ZAFIRA        VAN         1 2007-01-01
53859  2007    1  CHEVROLET  SERIE FTR     CAMION         1 2007-01-01
53858  2007    1  CHEVROLET  SERIE FTR     CAMION         1 2007-01-01
anio        0
mes         0
marca       0
familia     0
segmento    0
unidades    0
fecha       0
dtype: int64


In [13]:
# Eliminar filas sin unidades
df = df.dropna(subset=['unidades'])

# Crear variable de clasificación: alta vs baja venta
umbral = df['unidades'].median()
df['alta_venta'] = (df['unidades'] >= umbral).astype(int)

# Codificar categóricas
cat_cols = ['marca', 'familia', 'segmento']
df_encoded = pd.get_dummies(df, columns=cat_cols, drop_first=True)

# Definir features y targets
X = df_encoded.drop(columns=['unidades', 'alta_venta', 'fecha'])
y_clf = df_encoded['alta_venta']
y_reg = df_encoded['unidades']

print(X.isna().sum().sort_values(ascending=False).head())


segmento_VAN             0
familia_D-MAX            0
familia_SERIE B7R BUS    0
familia_SCROSS           0
familia_SAIL             0
dtype: int64


In [14]:
#Verificamos que X ya no tiene NaN.
X = X.fillna(0)


In [15]:
# Modelo de clasificación RandomForest Classifier: Predicción continua Para estimar unidades futuras por marca, 
#crecimiento proyectado, comportamiento esperado hasta 2029.

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

X_train, X_test, y_train, y_test = train_test_split(X, y_clf, test_size=0.2, random_state=42)

clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

y_pred_clf = clf.predict(X_test)
print(classification_report(y_test, y_pred_clf))


              precision    recall  f1-score   support

           0       0.53      0.24      0.33      4954
           1       0.68      0.88      0.77      9263

    accuracy                           0.66     14217
   macro avg       0.61      0.56      0.55     14217
weighted avg       0.63      0.66      0.62     14217



### Interpretacion del RandomForestClasifier: 

#### A. El modelo está favoreciendo fuertemente la clase 1
Precision clase 1: 0.68

Recall clase 1: 0.88

F1 clase 1: 0.77

El modelo detecta muy bien la clase 1 (crecimiento / desempeño positivo).

#### B. La clase 0 está siendo difícil de identificar
Precision clase 0: 0.53

Recall clase 0: 0.24

F1 clase 0: 0.33

El modelo casi no detecta la clase 0 (estancamiento / caída).

#### C. ¿Por qué ocurre esto?
Porque tu dataset está desbalanceado:

Clase 1: 9263 registros

Clase 0: 4954 registros

El modelo aprende:

“Es más seguro predecir clase 1 porque aparece más veces”.

Esto es típico en ventas de vehículos: la mayoría de marcas crecen, pocas caen.

#### D. Accuracy = 0.66
El modelo acierta 66% de las veces.

Pero el accuracy no es la mejor métrica aquí porque el dataset está desbalanceado.

E. Macro F1 = 0.55
Esta es la métrica más honesta:

Promedia ambas clases por igual.

Indica que el modelo tiene dificultad para distinguir comportamientos negativos.

#### Entonces:
El modelo RandomForestClassifier presenta un desempeño adecuado para identificar marcas con crecimiento (clase 1), alcanzando un recall del 88%. Sin embargo, la detección de marcas con comportamiento negativo (clase 0) es limitada debido al desbalance natural del dataset, donde la mayoría de marcas muestran crecimiento. El accuracy global del 66% y el macro F1 de 0.55 reflejan que el modelo es útil para clasificar tendencias positivas, pero menos preciso para identificar caídas o estancamientos.


In [16]:
# Modelo de regresión RandomForest Regressor: Para categorizar marcas según crecimiento, estabilidad,
# decrecimiento, desempeño comercial.

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Separar datos de entrenamiento y prueba
X_train_r, X_test_r, y_train_r, y_test_r = train_test_split(X, y_reg, test_size=0.2, random_state=42)

# Crear y entrenar el modelo
reg = RandomForestRegressor(random_state=42)
reg.fit(X_train_r, y_train_r)

# Predicciones
y_pred_r = reg.predict(X_test_r)

# Evaluación del modelo
mae = mean_absolute_error(y_test_r, y_pred_r)
rmse = np.sqrt(mean_squared_error(y_test_r, y_pred_r))
r2 = r2_score(y_test_r, y_pred_r)

print(f"MAE: {mae:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"R²: {r2:.2f}")



MAE: 10.45
RMSE: 24.35
R²: -0.01


### Interpretación del RandomForestRegressor
#### A. MAE = 10.45
En promedio, el modelo se equivoca por 10 unidades.

Si las ventas típicas son: 50, 100, 200, 500

un error de 10 unidades puede ser: aceptable para marcas grandes, significativo para marcas pequeñas.

#### B. RMSE = 24.35
El error cuadrático medio es mayor, lo que indica:

hay outliers,

hay marcas con variaciones fuertes,

el modelo sufre cuando las ventas cambian abruptamente.

#### C. R² = -0.01
Este es el punto crítico.

Un R² negativo significa: El modelo predice peor que una línea horizontal usando el promedio.

#### Es decir:

el modelo no está capturando la estructura de la variable, las ventas de vehículos son muy volátiles,

las variables explicativas no son suficientes para predecir con precisión. Esto es normal en datasets de ventas de vehículos:

hay shocks económicos,cambios de política,lanzamientos de modelos,discontinuaciones,promociones,estacionalidad.

Random Forest no es un modelo de series temporales, y por eso su R² puede ser bajo.

#### Conclusión del regresor (texto para tu TFM)
El modelo RandomForestRegressor presenta un MAE de 10.45 unidades y un RMSE de 24.35, lo que indica errores moderados en la predicción de ventas. Sin embargo, el valor de R² = -0.01 evidencia que el modelo no logra capturar adecuadamente la estructura temporal de las ventas de vehículos, probablemente debido a la alta variabilidad del mercado y a la ausencia de variables exógenas que expliquen cambios abruptos. Este resultado es coherente con la naturaleza del dataset y confirma que las ventas de vehículos presentan dinámicas complejas que no se ajustan bien a modelos de regresión basados únicamente en variables históricas.
